File: usememory/map/applied/sliding_window/longest_non_repeating_substring.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        // naive solution
        // for loop of all possible starting positions and 
        // for loop of all possible finishing positions
        // for o(n^2) complexity
        
        // maintain a rolling set of values in the substring.
        // i.e. start_index = 0, end_index = 0
        // loop through whilst finding chars not in set
        // if already in set then jump start from where that value was in set
        // therefore should use a map to know where to jump to
        Map<Character, Integer> substringMap = new HashMap<>();
        
        int substringStart = 0;
        int substringFinish = 0;
        int longestSubstring = 0;
        while(substringFinish < s.length()){
            char c = s.charAt(substringFinish);
            // c already exists so we jump substringStart to where that last 
            // instance was
            if(substringMap.containsKey(c)){
                int lastCIndex = substringMap.get(c);
                while(substringStart <= lastCIndex){
                    char k = s.charAt(substringStart);
                    substringMap.remove(k);
                    substringStart++;
                }
            }
            else {
                // can only possibly be longer if didn't have to remove from map
                if(substringFinish - substringStart >= longestSubstring){
                    longestSubstring = substringFinish - substringStart + 1;
                }
            }
            substringMap.put(c, substringFinish);
            substringFinish++;
        }
        return longestSubstring;
    }
}

public Solution{
    public int lengthOfLongestSubstring(String s){
        int longestSubstring = 0;
        Map<Character, Integer> substringMap = new HashMap<>();

        for(int substringStart = 0, substringFinish = 0; substringFinish < s.length(); substringFinish++){
            char c = s.charAt(substringFinish);
            if(substringMap.contains(c)){
                substringStart = Math.max(substringMap.get(c), substringStart);
            }
            longestSubstring = Math.max(longestSubstring, substringFinish - substringStart + 1);
            substringMap.put(c, substringFinish + 1);
        }
        return longestSubstring;
    }
}

File: two_pointers/map/longest_non_repeating_substring.go
package data_structures

func lengthOfLongestSubstring(s string) int {
	max, m := 0, map[byte]int{} // golang has no char so could either use byte or rune here
	for substringStart, substringFinish := 0, 0; substringFinish < len(s); substringFinish++ {
		// if map already has character at substringFinish increment substringStart
		if j, ok := m[s[substringFinish]]; ok && j >= substringStart {
			substringStart = j + 1 // could jump to substringStart max (k + 1) but math.Max only uses float
		}
		m[s[substringFinish]] = substringFinish
		// if new substring is longer then best so far update
		if l := substringFinish - substringStart + 1; l > max {
			max = l
		}
	}
	return max
}
